README.md
[![Build Status](https://api.travis-ci.com/danitso/terraform-provider-sftp.svg?branch=master)](https://travis-ci.com/danitso/terraform-provider-sftp)
[![Go Report Card](https://goreportcard.com/badge/github.com/danitso/terraform-provider-sftp)](https://goreportcard.com/report/github.com/danitso/terraform-provider-sftp)
[![GoDoc](https://godoc.org/github.com/danitso/terraform-provider-sftp?status.svg)](http://godoc.org/github.com/danitso/terraform-provider-sftp)

# Terraform Provider for SFTP
A Terraform Provider which adds additional SFTP functionality.

## Requirements

- [Terraform](https://www.terraform.io/downloads.html) 0.13+
- [Go](https://golang.org/doc/install) 1.15+ (to build the provider plugin)
- [GoReleaser](https://goreleaser.com/install/) 0.155+ (to build the provider plugin)

## Table of Contents
- [Building the provider](#building-the-provider)
- [Using the provider](#using-the-provider)
- [Testing the provider](#testing-the-provider)

## Building the provider
- Clone the repository to `$GOPATH/src/github.com/danitso/terraform-provider-sftp`:

    ```sh
    $ mkdir -p "${GOPATH}/src/github.com/danitso"
    $ cd "${GOPATH}/src/github.com/danitso"
    $ git clone git@github.com:danitso/terraform-provider-sftp
    ```

- Enter the provider directory and build it:

    ```sh
    $ cd "${GOPATH}/src/github.com/danitso/terraform-provider-sftp"
    $ make build
    ```

## Using the provider
You can find the latest release and its documentation in the [Terraform Registry](https://registry.terraform.io/providers/danitso/sftp/latest).

## Testing the provider
In order to test the provider, you can simply run `make test`.

```sh
$ make test
```

Tests are limited to regression tests, ensuring backwards compability.


CHANGELOG.md
## v0.2.0

ENHANCEMENTS:

* provider: Upgrade to Go v1.15
* provider: Upgrade dependencies

OTHER:

* provider/docs: Add HTML documentation powered by GitHub Pages and Terraform Registry

## 0.1.1

ENHANCEMENTS:

* provider: Upgraded to Go v1.13 and Terraform v0.12.18

## 0.1.0

FEATURES:

* **New Data Source:** `sftp_remote_file`


provider_test.go
package main

import (
	"testing"
)

// TestProviderInstantiation() tests whether the Provider instance can be instantiated.
func TestProviderInstantiation(t *testing.T) {
	s := Provider()

	if s == nil {
		t.Fatalf("Cannot instantiate Provider")
	}
}


resource_file_test.go
package main

import (
	"testing"
)

// TestResourceFileInstantiation tests whether the ResourceFile instance can be instantiated.
func TestResourceFileInstantiation(t *testing.T) {
	s := resourceFile()

	if s == nil {
		t.Fatalf("Cannot instantiate ResourceFile")
	}
}

// TestResourceFileSchema tests the ResourceFile schema.
func TestResourceFileSchema(t *testing.T) {
	s := resourceFile()

	requiredKeys := []string{
		dataSourceDiskLabelKey,
		dataSourceDiskServerIDKey,
		dataSourceDiskSizeKey,
	}

	for _, v := range requiredKeys {
		if s.Schema[v] == nil {
			t.Fatalf("Error in ResourceFile.Schema: Missing argument \"%s\"", v)
		}

		if s.Schema[v].Required != true {
			t.Fatalf("Error in ResourceFile.Schema: Argument \"%s\" is not required", v)
		}
	}

	attributeKeys := []string{
		dataSourceDiskPrimaryKey,
	}

	for _, v := range attributeKeys {
		if s.Schema[v] == nil {
			t.Fatalf("Error in dataSourceDisk.Schema: Missing attribute \"%s\"", v)
		}

		if s.Schema[v].Computed != true {
			t.Fatalf("Error in dataSourceDisk.Schema: Attribute \"%s\" is not computed", v)
		}
	}
}


sftptf/provider_test.go
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

package sftptf

import (
	"testing"
)

// TestProviderInstantiation() tests whether the Provider instance can be instantiated.
func TestProviderInstantiation(t *testing.T) {
	s := Provider()

	if s == nil {
		t.Fatalf("Cannot instantiate Provider")
	}
}


sftptf/provider.go
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

package sftptf

import (
	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
)

// Provider returns the object for this provider.
func Provider() *schema.Provider {
	return &schema.Provider{
		ConfigureFunc: providerConfigure,
		DataSourcesMap: map[string]*schema.Resource{
			"sftp_remote_file": dataSourceRemoteFile(),
		},
		ResourcesMap: map[string]*schema.Resource{},
		Schema:       map[string]*schema.Schema{},
	}
}

// providerConfigure() configures the provider before processing any IronMQ resources.
func providerConfigure(d *schema.ResourceData) (interface{}, error) {
	return nil, nil
}


sftptf/data_source_remote_file.go
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

package sftptf

import (
	"bytes"
	"errors"
	"fmt"
	"io"
	"time"

	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
	"github.com/pkg/sftp"
	"golang.org/x/crypto/ssh"
)

const (
	mkDataSourceRemoteFileAllowMissing = "allow_missing"
	mkDataSourceRemoteFileContents     = "contents"
	mkDataSourceRemoteFileHost         = "host"
	mkDataSourceRemoteFileHostKey      = "host_key"
	mkDataSourceRemoteFileLastModified = "last_modified"
	mkDataSourceRemoteFilePassword     = "password"
	mkDataSourceRemoteFilePath         = "path"
	mkDataSourceRemoteFilePort         = "port"
	mkDataSourceRemoteFilePrivateKey   = "private_key"
	mkDataSourceRemoteFileSize         = "size"
	mkDataSourceRemoteFileTimeout      = "timeout"
	mkDataSourceRemoteFileTriggers     = "triggers"
	mkDataSourceRemoteFileUser         = "user"
)

// dataSourceRemoteFile retrieves information about a remote file.
func dataSourceRemoteFile() *schema.Resource {
	return &schema.Resource{
		Schema: map[string]*schema.Schema{
			mkDataSourceRemoteFileAllowMissing: {
				Type:        schema.TypeBool,
				Description: "Whether to ignore that the file is missing",
				Optional:    true,
				Default:     false,
				ForceNew:    true,
			},
			mkDataSourceRemoteFileContents: {
				Type:        schema.TypeString,
				Description: "The file contents",
				Computed:    true,
				ForceNew:    true,
			},
			mkDataSourceRemoteFileHost: {
				Type:        schema.TypeString,
				Description: "The hostname",
				Required:    true,
				ForceNew:    true,
			},
			mkDataSourceRemoteFileHostKey: {
				Type:        schema.TypeString,
				Description: "The host key",
				Optional:    true,
				Default:     "",
				ForceNew:    true,
			},
			mkDataSourceRemoteFileLastModified: {
				Type:        schema.TypeString,
				Description: "The last modified timestamp",
				Computed:    true,
				ForceNew:    true,
			},
			mkDataSourceRemoteFilePassword: {
				Type:        schema.TypeString,
				Description: "The password",
				Optional:    true,
				Default:     "",
				ForceNew:    true,
			},
			mkDataSourceRemoteFilePath: {
				Type:        schema.TypeString,
				Description: "The file path",
				Required:    true,
				ForceNew:    true,
			},
			mkDataSourceRemoteFilePort: {
				Type:        schema.TypeInt,
				Description: "The port number",
				Optional:    true,
				Default:     22,
				ForceNew:    true,
			},
			mkDataSourceRemoteFilePrivateKey: {
				Type:        schema.TypeString,
				Description: "The private key",
				Optional:    true,
				Default:     "",
				ForceNew:    true,
			},
			mkDataSourceRemoteFileSize: {
				Type:        schema.TypeInt,
				Description: "The file size (in bytes)",
				Computed:    true,
				ForceNew:    true,
			},
			mkDataSourceRemoteFileTimeout: {
				Type:        schema.TypeString,
				Description: "The connect timeout",
				Optional:    true,
				Default:     "5m",
				ForceNew:    true,
			},
			mkDataSourceRemoteFileTriggers: {
				Type:        schema.TypeMap,
				Description: "The triggers",
				Optional:    true,
				ForceNew:    true,
			},
			mkDataSourceRemoteFileUser: {
				Type:        schema.TypeString,
				Description: "The username",
				Optional:    true,
				Default:     "",
				ForceNew:    true,
			},
		},

		Read: dataSourceRemoteFileRead,
	}
}

// dataSourceRemoteFileCreateSSHClient creates a new SSH client.
func dataSourceRemoteFileCreateSSHClient(d *schema.ResourceData, m interface{}) (*ssh.Client, error) {
	host := d.Get(mkDataSourceRemoteFileHost).(string)
	hostKey := d.Get(mkDataSourceRemoteFileHostKey).(string)
	password := d.Get(mkDataSourceRemoteFilePassword).(string)
	port := d.Get(mkDataSourceRemoteFilePort).(int)
	privateKey := d.Get(mkDataSourceRemoteFilePrivateKey).(string)
	timeout, err := time.ParseDuration(d.Get(mkDataSourceRemoteFileTimeout).(string))
	username := d.Get(mkDataSourceRemoteFileUser).(string)

	if err != nil {
		return nil, err
	}

	if password == "" && privateKey == "" {
		return nil, errors.New("No password or private key has been specified")
	}

	var authMethod []ssh.AuthMethod

	if password != "" {
		authMethod = []ssh.AuthMethod{ssh.Password(password)}
	} else {
		privateKeySigner, err := ssh.ParsePrivateKey([]byte(privateKey))

		if err != nil {
			return nil, err
		}

		authMethod = []ssh.AuthMethod{ssh.PublicKeys(privateKeySigner)}
	}

	var hostKeyCallback ssh.HostKeyCallback

	if hostKey != "" {
		parsedHostKey, err := ssh.ParsePublicKey([]byte(hostKey))

		if err != nil {
			return nil, err
		}

		hostKeyCallback = ssh.FixedHostKey(parsedHostKey)
	} else {
		hostKeyCallback = ssh.InsecureIgnoreHostKey()
	}

	sshConfig := &ssh.ClientConfig{
		User:            username,
		Auth:            authMethod,
		HostKeyCallback: hostKeyCallback,
	}

	timeDelay := int64(10)
	timeMax := timeout.Seconds()
	timeStart := time.Now()
	timeElapsed := timeStart.Sub(timeStart)

	err = nil

	var client *ssh.Client

	for timeElapsed.Seconds() < timeMax {
		if int64(timeElapsed.Seconds())%timeDelay == 0 {
			client, err = ssh.Dial("tcp", fmt.Sprintf("%s:%d", host, port), sshConfig)

			if err == nil {
				break
			}

			time.Sleep(1 * time.Second)
		}

		time.Sleep(200 * time.Millisecond)

		timeElapsed = time.Now().Sub(timeStart)
	}

	if err != nil {
		return nil, err
	}

	return client, nil
}

// dataSourceRemoteFileRead reads information about a remote file.
func dataSourceRemoteFileRead(d *schema.ResourceData, m interface{}) error {
	allowMissing := d.Get(mkDataSourceRemoteFileAllowMissing).(bool)
	remoteFilePath := d.Get(mkDataSourceRemoteFilePath).(string)

	// Create a new SFTP client.
	sshClient, err := dataSourceRemoteFileCreateSSHClient(d, m)

	if err != nil {
		return err
	}

	defer sshClient.Close()

	sftpClient, err := sftp.NewClient(sshClient)

	if err != nil {
		return err
	}

	defer sftpClient.Close()

	// Retrieve the information for a remote file as well as its data.
	remoteFileInfo, err := sftpClient.Lstat(remoteFilePath)

	if err != nil {
		if allowMissing {
			d.SetId("missing")

			d.Set(mkDataSourceRemoteFileContents, "")
			d.Set(mkDataSourceRemoteFileLastModified, time.Now().Format(time.RFC3339))
			d.Set(mkDataSourceRemoteFileSize, -1)

			return nil
		}

		return err
	}

	remoteFile, err := sftpClient.Open(remoteFilePath)

	if err != nil {
		return err
	}

	defer remoteFile.Close()

	buffer := bytes.NewBuffer(nil)
	_, err = io.Copy(buffer, remoteFile)

	if err != nil {
		return err
	}

	d.SetId(remoteFileInfo.Name())

	d.Set(mkDataSourceRemoteFileContents, buffer.String())
	d.Set(mkDataSourceRemoteFileLastModified, remoteFileInfo.ModTime().Format(time.RFC3339))
	d.Set(mkDataSourceRemoteFileSize, int(remoteFileInfo.Size()))

	return nil
}


resource_file.go
package main

import (
	"bytes"
	"errors"
	"fmt"
	"io"
	"os"
	"strconv"
	"time"

	"github.com/hashicorp/terraform/helper/schema"
	"github.com/pkg/sftp"
	"golang.org/x/crypto/ssh"
)

const (
	mkResourceFileContents          = "contents"
	mkResourceFileDestroyLocalFile  = "destroy_local_file"
	mkResourceFileDestroyRemoteFile = "destroy_remote_file"
	mkResourceFileDownload          = "download"
	mkResourceFileHost              = "host"
	mkResourceFileHostKey           = "host_key"
	mkResourceFileLastModified      = "last_modified"
	mkResourceFileLocalFilePath     = "local_file_path"
	mkResourceFilePassword          = "password"
	mkResourceFilePort              = "port"
	mkResourceFilePrivateKey        = "private_key"
	mkResourceFileRemoteFilePath    = "remote_file_path"
	mkResourceFileSize              = "size"
	mkResourceFileTimeout           = "timeout"
	mkResourceFileTriggers          = "triggers"
	mkResourceFileUser              = "user"
)

// resourceFile manages a file.
func resourceFile() *schema.Resource {
	return &schema.Resource{
		Schema: map[string]*schema.Schema{
			mkResourceFileContents: &schema.Schema{
				Type:        schema.TypeString,
				Description: "The file contents",
				Optional:    true,
				ForceNew:    true,
			},
			mkResourceFileDestroyLocalFile: &schema.Schema{
				Type:        schema.TypeBool,
				Description: "Whether to destroy the local file when the resource is destroyed",
				Optional:    true,
				Default:     false,
			},
			mkResourceFileDestroyRemoteFile: &schema.Schema{
				Type:        schema.TypeBool,
				Description: "Whether to destroy the remote file when the resource is destroyed",
				Optional:    true,
				Default:     false,
			},
			mkResourceFileDownload: &schema.Schema{
				Type:        schema.TypeBool,
				Description: "Whether to download the file",
				Optional:    true,
				Default:     true,
			},
			mkResourceFileHost: &schema.Schema{
				Type:        schema.TypeString,
				Description: "The hostname",
				Required:    true,
				ForceNew:    true,
			},
			mkResourceFileHostKey: &schema.Schema{
				Type:        schema.TypeString,
				Description: "The host key",
				Optional:    true,
				Default:     "",
			},
			mkResourceFileLastModified: &schema.Schema{
				Type:        schema.TypeString,
				Description: "The last modified timestamp",
				Computed:    true,
				ForceNew:    true,
			},
			mkResourceFileLocalFilePath: &schema.Schema{
				Type:        schema.TypeString,
				Description: "The local file path",
				Optional:    true,
				Default:     "",
				ForceNew:    true,
			},
			mkResourceFilePassword: &schema.Schema{
				Type:        schema.TypeString,
				Description: "The password",
				Optional:    true,
				Default:     "",
			},
			mkResourceFilePort: &schema.Schema{
				Type:        schema.TypeInt,
				Description: "The port number",
				Optional:    true,
				Default:     22,
			},
			mkResourceFilePrivateKey: &schema.Schema{
				Type:        schema.TypeString,
				Description: "The private key",
				Optional:    true,
				Default:     "",
			},
			mkResourceFileRemoteFilePath: &schema.Schema{
				Type:        schema.TypeString,
				Description: "The remote file path",
				Required:    true,
				ForceNew:    true,
			},
			mkResourceFileSize: &schema.Schema{
				Type:        schema.TypeString,
				Description: "The file size in bytes",
				Computed:    true,
				ForceNew:    true,
			},
			mkResourceFileTimeout: &schema.Schema{
				Type:        schema.TypeString,
				Description: "The connection timeout",
				Optional:    true,
				Default:     "5m",
			},
			mkResourceFileTriggers: &schema.Schema{
				Type:     schema.TypeMap,
				Optional: true,
				ForceNew: true,
			},
			mkResourceFileUser: &schema.Schema{
				Type:        schema.TypeString,
				Description: "The username",
				Optional:    true,
				Default:     "",
			},
		},

		Create: resourceFileCreate,
		Read:   resourceFileRead,
		Delete: resourceFileDelete,
	}
}

// resourceFileCreate initializes a file download or upload.
func resourceFileCreate(d *schema.ResourceData, m interface{}) error {
	download := d.Get(mkResourceFileDownload).(bool)
	localFilePath := d.Get(mkResourceFileLocalFilePath).(string)
	remoteFilePath := d.Get(mkResourceFileRemoteFilePath).(string)

	// Create a new SFTP client.
	sshClient, err := resourceFileCreateSSHClient(d, m)

	if err != nil {
		return err
	}

	defer sshClient.Close()

	sftpClient, err := sftp.NewClient(sshClient)

	if err != nil {
		return err
	}

	defer sftpClient.Close()

	// Download or upload the file.
	var remoteFile *sftp.File

	if download {
		remoteFile, err = sftpClient.Open(remoteFilePath)
	} else {
		remoteFile, err = sftpClient.Create(remoteFilePath)
	}

	if err != nil {
		return err
	}

	defer remoteFile.Close()

	if localFilePath != "" {
		localFile, err := os.OpenFile(localFilePath, os.O_RDWR, 0644)

		if err != nil {
			return err
		}

		defer localFile.Close()

		if download {
			_, err = io.Copy(localFile, remoteFile)
		} else {
			_, err = io.Copy(remoteFile, localFile)
		}

		if err != nil {
			return err
		}
	} else {
		if download {
			buffer := bytes.NewBuffer(nil)
			_, err = io.Copy(buffer, remoteFile)

			d.Set(mkResourceFileContents, buffer.String())
		} else {
			buffer := bytes.NewBufferString(d.Get(mkResourceFileContents).(string))
			_, err = io.Copy(remoteFile, buffer)
		}

		if err != nil {
			return err
		}
	}

	// Retrieve information about the file.
	remoteFileInfo, err := sftpClient.Lstat(remoteFilePath)

	if err != nil {
		return err
	}

	d.SetId(remoteFileInfo.Name())

	d.Set(mkResourceFileLastModified, remoteFileInfo.ModTime().Format(time.RFC3339))
	d.Set(mkResourceFileSize, strconv.FormatInt(remoteFileInfo.Size(), 10))

	return nil
}

// resourceFileCreateSSHClient creates a new SSH client.
func resourceFileCreateSSHClient(d *schema.ResourceData, m interface{}) (*ssh.Client, error) {
	host := d.Get(mkResourceFileHost).(string)
	hostKey := d.Get(mkResourceFileHostKey).(string)
	password := d.Get(mkResourceFilePassword).(string)
	port := d.Get(mkResourceFilePort).(int)
	privateKey := d.Get(mkResourceFilePrivateKey).(string)
	timeout, err := time.ParseDuration(d.Get(mkResourceFileTimeout).(string))
	username := d.Get(mkResourceFileUser).(string)

	if err != nil {
		return nil, err
	}

	if password == "" && privateKey == "" {
		return nil, errors.New("No password or private key has been specified")
	}

	var authMethod []ssh.AuthMethod

	if password != "" {
		authMethod = []ssh.AuthMethod{ssh.Password(password)}
	} else {
		privateKeySigner, err := ssh.ParsePrivateKey([]byte(privateKey))

		if err != nil {
			return nil, err
		}

		authMethod = []ssh.AuthMethod{ssh.PublicKeys(privateKeySigner)}
	}

	var hostKeyCallback ssh.HostKeyCallback

	if hostKey != "" {
		parsedHostKey, err := ssh.ParsePublicKey([]byte(hostKey))

		if err != nil {
			return nil, err
		}

		hostKeyCallback = ssh.FixedHostKey(parsedHostKey)
	} else {
		hostKeyCallback = ssh.InsecureIgnoreHostKey()
	}

	sshConfig := &ssh.ClientConfig{
		User:            username,
		Auth:            authMethod,
		HostKeyCallback: hostKeyCallback,
	}

	timeDelay := int64(10)
	timeMax := timeout.Seconds()
	timeStart := time.Now()
	timeElapsed := timeStart.Sub(timeStart)

	err = nil

	var client *ssh.Client

	for timeElapsed.Seconds() < timeMax {
		if int64(timeElapsed.Seconds())%timeDelay == 0 {
			client, err = ssh.Dial("tcp", fmt.Sprintf("%s:%d", host, port), sshConfig)

			if err == nil {
				break
			}

			time.Sleep(1 * time.Second)
		}

		time.Sleep(200 * time.Millisecond)

		timeElapsed = time.Now().Sub(timeStart)
	}

	if err != nil {
		return nil, err
	}

	return client, nil
}

// resourceFileRead reads information about a remote file.
func resourceFileRead(d *schema.ResourceData, m interface{}) error {
	download := d.Get(mkResourceFileDownload).(bool)
	localFilePath := d.Get(mkResourceFileLocalFilePath).(string)
	remoteFilePath := d.Get(mkResourceFileRemoteFilePath).(string)

	// Create a new SFTP client.
	sshClient, err := resourceFileCreateSSHClient(d, m)

	if err != nil {
		return err
	}

	defer sshClient.Close()

	sftpClient, err := sftp.NewClient(sshClient)

	if err != nil {
		return err
	}

	defer sftpClient.Close()

	// Retrieve information about the files in order to determine, if the resource needs to be re-created.
	remoteFileInfo, err := sftpClient.Lstat(remoteFilePath)

	if err != nil {
		if !download {
			d.SetId("")

			return nil
		}

		return err
	}

	if localFilePath != "" {
		localFileInfo, err := os.Stat(localFilePath)

		if err != nil {
			if os.IsNotExist(err) && download {
				d.SetId("")

				return nil
			}

			return err
		}

		if localFileInfo.Size() != remoteFileInfo.Size() {
			d.SetId("")

			return nil
		}
	} else {
		contents := d.Get(mkResourceFileContents).(string)

		if int64(len(contents)) != remoteFileInfo.Size() {
			d.SetId("")

			return nil
		}
	}

	d.Set(mkResourceFileLastModified, remoteFileInfo.ModTime().Format(time.RFC3339))
	d.Set(mkResourceFileSize, strconv.FormatInt(remoteFileInfo.Size(), 10))

	return nil
}

// resourceFileDelete cleans up after a file download or upload.
func resourceFileDelete(d *schema.ResourceData, m interface{}) error {
	if d.Get(mkResourceFileDestroyLocalFile).(bool) {
		localFilePath := d.Get(mkResourceFileLocalFilePath).(string)

		if localFilePath != "" {
			if _, err := os.Stat(localFilePath); err == nil {
				err := os.Remove(localFilePath)

				if err != nil {
					return err
				}
			}
		}
	}

	if d.Get(mkResourceFileDestroyRemoteFile).(bool) {
		remoteFilePath := d.Get(mkResourceFileRemoteFilePath).(string)

		// Create a new SFTP client and delete the remote file.
		sshClient, err := resourceFileCreateSSHClient(d, m)

		if err != nil {
			return err
		}

		defer sshClient.Close()

		sftpClient, err := sftp.NewClient(sshClient)

		if err != nil {
			return err
		}

		defer sftpClient.Close()

		if _, err := sftpClient.Lstat(remoteFilePath); err == nil {
			err = sftpClient.Remove(remoteFilePath)

			if err != nil {
				return err
			}
		}
	}

	d.SetId("")

	return nil
}
